ouroboros-consensus/src/consensus-testlib/Test/Ouroboros/Consensus/ChainGenerator/Adversarial.hs

```diff
@@ -0,0 +1,525 @@
+-- dense chain would have to lose at least one race within the Genesis window.
+checkAdversarialChain ::
+  forall base hon adv.
+     AdversarialRecipe base hon
+  -> ChainSchema base adv
+  -> Exn.Except (AdversarialViolation hon adv) ()
+checkAdversarialChain recipe adv = do
+    checkStart
+    checkCount
+    checkRaces
+  where
+    AdversarialRecipe {
+        arHonest = ChainSchema winH vH
+      ,
+        arParams = (Kcp k, Scg s, Delta d)
+      ,
+        arPrefix
+      } = recipe
+
+    ChainSchema winA vA = adv
+
+    checkStart = do
+        let startA       = C.windowStart winA :: C.Index base SlotE
+            intersection = startA C.- 1       :: C.Index base SlotE
+
+        case C.toWin winH intersection of
+            Nothing -> do
+              -- genesis block is the only permissible anchor outside of @hon@
+              when (startA   /= C.Count 0) $ Exn.throwError $ BadAnchor HonestActiveMustAnchorAdversarial
+              when (arPrefix /= C.Count 0) $ Exn.throwError $ BadAnchor WrongNumberOfHonestPredecessors
+
+            Just i -> do
+                let _ = i :: C.Index hon SlotE
+
+                when (BV.testV S.inverted vH i) $ do
+                    Exn.throwError $ BadAnchor HonestActiveMustAnchorAdversarial
+
+                C.SomeWindow Proxy precedingSlots <-
+                    pure $ C.withWindowBetween (C.windowSize winH) (C.Lbl @"foo") (C.Count 0) i
+                let pc = BV.countActivesInV S.notInverted (C.sliceV precedingSlots vH)
+
+                when (C.frWinVar precedingSlots (C.toVar pc) /= arPrefix) $ do
+                    Exn.throwError $ BadAnchor WrongNumberOfHonestPredecessors
+
+    checkCount = do
+        let pc = BV.countActivesInV S.notInverted vA
+        when (C.toVar pc <= 0) $ Exn.throwError BadCount
+
+    -- the youngest slot in the adversarial schedule that cannot have accelerated
+    --
+    -- (IE s past the first active adversarial slot.)
+    youngestStableA :: C.Index adv SlotE
+    youngestStableA =
+        case BV.findIthEmptyInV S.inverted vA (C.Count 0) of
+            BV.JustFound firstActiveA ->
+                -- if s=0, then the slot of their block is the youngest stable slot
+                firstActiveA C.+ s
+            BV.NothingFound           ->
+                -- the rest of the function won' force this since there are no
+                -- adverarial active slots
+                error "dead code"
+
+    checkRaces = do
+        let iterH =
+                maybe (RI.initConservative (Scg s) (Delta d) winH) id
+              $ RI.init (Kcp k) vH
+        case RI.init (Kcp (k - 1)) vA >>= RI.next vA of
+            Nothing    -> pure ()   -- there are <= k total adversarial active slots
+            Just iterA ->
+                -- TODO optimization: how to skip all the honest Race Windows that
+                -- don't reach beyond the intersection? Perhaps skip to i - s + d?
+                go iterH iterA
```

Comment:
Addressing the TODO would require counting blocks backwards from the intersection. Is that ideal?